Mathlib/Lean/Exception.lean

```diff
@@ -0,0 +1,7 @@
+import Lean
+open Lean
+
+def successIfFail [MonadError M] [Monad M] (m : M α) : M Exception := do
+  match ← tryCatch (m *> pure none) (pure ∘ some) with
+  | none => throwError "Expected an exception."
+  | some ex => return ex
```

Comment:
I wonder if this is clearer:
```suggestion
  try
    discard m
    throwError "Expected an exception."
  catch ex =>
    return ex
```

Reply:
hmm I think this doesn't quite work because it needs to throw if m doesn't throw